File: packages/wasm-api/test/fixtures/string-slice.ts
// @ts-ignore possibly includes unused imports
import { MemorySlice, Pointer, WasmStringSlice, WasmTypeBase, WasmTypeConstructor } from "@thi.ng/wasm-api";

export interface Foo extends WasmTypeBase {
	single: WasmStringSlice;
	singleMut: WasmStringSlice;
	multi: WasmStringSlice[];
	singlePtr: Pointer<WasmStringSlice>;
	multiPtr: Pointer<WasmStringSlice[]>;
	slice: WasmStringSlice[];
	mutSlice: WasmStringSlice[];
}

export const $Foo: WasmTypeConstructor<Foo> = (mem) => ({
	get align() {
		return 4;
	},
	get size() {
		return 56;
	},
	instance: (base) => {
		let $singlePtr: Pointer<WasmStringSlice> | null = null;
		let $multiPtr: Pointer<WasmStringSlice[]> | null = null;
		let $single: WasmStringSlice | null = null;
		let $singleMut: WasmStringSlice | null = null;
		return {
			get __base() {
				return base;
			},
			get __bytes() {
				return mem.u8.subarray(base, base + 56);
			},
			get single(): WasmStringSlice {
				return $single || ($single = new WasmStringSlice(mem, base, true));
			},
			get singleMut(): WasmStringSlice {
				return $singleMut || ($singleMut = new WasmStringSlice(mem, (base + 8), false));
			},
			get multi(): WasmStringSlice[] {
				const addr = (base + 16);
				const $multi: WasmStringSlice[] = [];
				for(let i = 0; i < 2; i++) $multi.push(new WasmStringSlice(mem, addr + i * 8, true));
				return $multi;
			},
			get singlePtr(): Pointer<WasmStringSlice> {
				return $singlePtr || ($singlePtr = new Pointer<WasmStringSlice>(mem, (base + 32),
				(addr) => new WasmStringSlice(mem, addr, true)
				));
			},
			get multiPtr(): Pointer<WasmStringSlice[]> {
				return $multiPtr || ($multiPtr = new Pointer<WasmStringSlice[]>(mem, (base + 36),
				(addr) => {
					const $buf: WasmStringSlice[] = [];
					for(let i = 0; i < 2; i++) $buf.push(new WasmStringSlice(mem, addr + i * 8, true));
					return $buf;
				}
				));
			},
			get slice(): WasmStringSlice[] {
				const addr = mem.u32[(base + 40) >>> 2];
				const len = mem.u32[(base + 44) >>> 2];
				const $slice: WasmStringSlice[] = [];
				for(let i = 0; i < len; i++) $slice.push(new WasmStringSlice(mem, addr + i * 8, true));
				return $slice;
			},
			get mutSlice(): WasmStringSlice[] {
				const addr = mem.u32[(base + 48) >>> 2];
				const len = mem.u32[(base + 52) >>> 2];
				const $mutSlice: WasmStringSlice[] = [];
				for(let i = 0; i < len; i++) $mutSlice.push(new WasmStringSlice(mem, addr + i * 8, false));
				return $mutSlice;
			},
		};
	}
});

File: packages/wasm-api/test/fixtures/string-slice.zig
const std = @import("std");
const wasm = @import("wasmapi");

pub const Foo = extern struct {
    single: wasm.ConstString,
    singleMut: wasm.String,
    multi: [2]wasm.ConstString,
    singlePtr: *wasm.ConstString,
    multiPtr: *[2]wasm.ConstString,
    slice: wasm.ConstStringSlice,
    mutSlice: wasm.StringSlice,
};

export fn Foo_align() usize {
    return @alignOf(Foo);
}

export fn Foo_size() usize {
    return @sizeOf(Foo);
}

export fn Foo_single_align() usize {
    return @alignOf(wasm.ConstString);
}

export fn Foo_single_offset() usize {
    return @offsetOf(Foo, "single");
}

export fn Foo_single_size() usize {
    return @sizeOf(wasm.ConstString);
}

export fn Foo_singleMut_align() usize {
    return @alignOf(wasm.String);
}

export fn Foo_singleMut_offset() usize {
    return @offsetOf(Foo, "singleMut");
}

export fn Foo_singleMut_size() usize {
    return @sizeOf(wasm.String);
}

export fn Foo_multi_align() usize {
    return @alignOf([2]wasm.ConstString);
}

export fn Foo_multi_offset() usize {
    return @offsetOf(Foo, "multi");
}

export fn Foo_multi_size() usize {
    return @sizeOf([2]wasm.ConstString);
}

export fn Foo_singlePtr_align() usize {
    return @alignOf(*wasm.ConstString);
}

export fn Foo_singlePtr_offset() usize {
    return @offsetOf(Foo, "singlePtr");
}

export fn Foo_singlePtr_size() usize {
    return @sizeOf(*wasm.ConstString);
}

export fn Foo_multiPtr_align() usize {
    return @alignOf(*[2]wasm.ConstString);
}

export fn Foo_multiPtr_offset() usize {
    return @offsetOf(Foo, "multiPtr");
}

export fn Foo_multiPtr_size() usize {
    return @sizeOf(*[2]wasm.ConstString);
}

export fn Foo_slice_align() usize {
    return @alignOf(wasm.ConstStringSlice);
}

export fn Foo_slice_offset() usize {
    return @offsetOf(Foo, "slice");
}

export fn Foo_slice_size() usize {
    return @sizeOf(wasm.ConstStringSlice);
}

export fn Foo_mutSlice_align() usize {
    return @alignOf(wasm.StringSlice);
}

export fn Foo_mutSlice_offset() usize {
    return @offsetOf(Foo, "mutSlice");
}

export fn Foo_mutSlice_size() usize {
    return @sizeOf(wasm.StringSlice);
}
